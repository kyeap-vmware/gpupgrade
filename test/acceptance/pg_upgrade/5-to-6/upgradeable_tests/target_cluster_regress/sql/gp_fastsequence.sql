-- Copyright (c) 2017-2023 VMware, Inc. or its affiliates
-- SPDX-License-Identifier: Apache-2.0

--------------------------------------------------------------------------------
-- Validate gp_fastsequence was upgraded
--------------------------------------------------------------------------------

-- Verify table's gp_fastsequence value is preserved
SELECT fs.gp_segment_id, fs.objmod, fs.last_sequence
FROM pg_class c
JOIN pg_appendonly ao ON c.oid=ao.relid
JOIN gp_dist_random('gp_fastsequence') fs ON ao.segrelid=fs.objid
WHERE c.relname='aotable_fastsequence'
ORDER BY 1, 2, 3;

-- Verify table data is not corrupt using seqscan
SET enable_indexscan = false;
SET enable_bitmapscan = false;
SET enable_seqscan = true;
SET optimizer = off;

EXPLAIN (COSTS OFF) SELECT * FROM aotable_fastsequence ORDER BY i;
SELECT * FROM aotable_fastsequence ORDER BY i;

-- Verify INSERTs produce no duplicate ctids
1: BEGIN;
1: INSERT INTO aotable_fastsequence SELECT generate_series(1001, 1010);
2: INSERT INTO aotable_fastsequence SELECT generate_series(1011, 1020);
1: COMMIT;
SELECT gp_segment_id, ctid, count(ctid) FROM aotable_fastsequence GROUP BY gp_segment_id, ctid HAVING count(ctid) > 1;

SET enable_indexscan = true;
SET enable_bitmapscan = true;
SET enable_seqscan = false;

EXPLAIN (COSTS OFF) SELECT * FROM aotable_fastsequence ORDER BY i;
SELECT * FROM aotable_fastsequence WHERE i < 10 ORDER BY i;

-- Verify table's gp_fastsequence value is preserved
SELECT fs.gp_segment_id, fs.objmod, fs.last_sequence
FROM pg_class c
JOIN pg_appendonly ao ON c.oid=ao.relid
JOIN gp_dist_random('gp_fastsequence') fs ON ao.segrelid=fs.objid
WHERE c.relname='aocotable_fastsequence'
ORDER BY 1, 2, 3;

-- Verify table data is not corrupt using seqscan
SET enable_indexscan = false;
SET enable_bitmapscan = false;
SET enable_seqscan = true;

EXPLAIN (COSTS OFF) SELECT * FROM aocotable_fastsequence ORDER BY i;
SELECT * FROM aocotable_fastsequence ORDER BY i;

-- Verify INSERTs produce no duplicate ctids
1: BEGIN;
1: INSERT INTO aocotable_fastsequence SELECT generate_series(1001, 1010);
2: INSERT INTO aocotable_fastsequence SELECT generate_series(1011, 1020);
1: COMMIT;
SELECT gp_segment_id, ctid, count(ctid) FROM aocotable_fastsequence GROUP BY gp_segment_id, ctid HAVING count(ctid) > 1;

SET enable_indexscan = true;
SET enable_bitmapscan = true;
SET enable_seqscan = false;

EXPLAIN (COSTS OFF) SELECT * FROM aocotable_fastsequence ORDER BY i;
SELECT * FROM aocotable_fastsequence WHERE i < 10 ORDER BY i;
