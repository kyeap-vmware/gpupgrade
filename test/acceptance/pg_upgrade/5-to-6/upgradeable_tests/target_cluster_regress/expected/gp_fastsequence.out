-- Copyright (c) 2017-2023 VMware, Inc. or its affiliates
-- SPDX-License-Identifier: Apache-2.0

--------------------------------------------------------------------------------
-- Validate gp_fastsequence was upgraded
--------------------------------------------------------------------------------

-- Verify table's gp_fastsequence value is preserved
SELECT fs.gp_segment_id, fs.objmod, fs.last_sequence FROM pg_class c JOIN pg_appendonly ao ON c.oid=ao.relid JOIN gp_dist_random('gp_fastsequence') fs ON ao.segrelid=fs.objid WHERE c.relname='aotable_fastsequence' ORDER BY 1, 2, 3;
 gp_segment_id | objmod | last_sequence 
---------------+--------+---------------
 0             | 0      | 0             
 0             | 1      | 301           
 0             | 2      | 101           
 1             | 0      | 0             
 1             | 1      | 301           
 1             | 2      | 101           
 2             | 0      | 0             
 2             | 1      | 301           
 2             | 2      | 101           
(9 rows)

-- Verify table data is not corrupt using seqscan
SET enable_indexscan = false;
SET
SET enable_bitmapscan = false;
SET
SET enable_seqscan = true;
SET
SET optimizer = off;
SET

EXPLAIN (COSTS OFF) SELECT * FROM aotable_fastsequence ORDER BY i;
 QUERY PLAN                                   
----------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)     
   Merge Key: i                               
   ->  Sort                                   
         Sort Key: i                          
         ->  Seq Scan on aotable_fastsequence 
 Optimizer: Postgres query optimizer          
(6 rows)
SELECT * FROM aotable_fastsequence ORDER BY i;
 i   
-----
 1   
 2   
 3   
 4   
 5   
 6   
 7   
 8   
 9   
 10  
 11  
 12  
 13  
 14  
 15  
 16  
 17  
 18  
 19  
 20  
 21  
 22  
 23  
 24  
 25  
 26  
 27  
 28  
 29  
 30  
 102 
 103 
 104 
 105 
 106 
 107 
 108 
 109 
 110 
 111 
 112 
 113 
 114 
 115 
 116 
 117 
 118 
 119 
 120 
 121 
(50 rows)

-- Verify INSERTs produce no duplicate ctids
1: BEGIN;
BEGIN
1: INSERT INTO aotable_fastsequence SELECT generate_series(1001, 1010);
INSERT 10
2: INSERT INTO aotable_fastsequence SELECT generate_series(1011, 1020);
INSERT 10
1: COMMIT;
COMMIT
SELECT gp_segment_id, ctid, count(ctid) FROM aotable_fastsequence GROUP BY gp_segment_id, ctid HAVING count(ctid) > 1;
 gp_segment_id | ctid | count 
---------------+------+-------
(0 rows)

SET enable_indexscan = true;
SET
SET enable_bitmapscan = true;
SET
SET enable_seqscan = false;
SET

EXPLAIN (COSTS OFF) SELECT * FROM aotable_fastsequence ORDER BY i;
 QUERY PLAN                                                      
-----------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)                        
   Merge Key: i                                                  
   ->  Sort                                                      
         Sort Key: i                                             
         ->  Bitmap Heap Scan on aotable_fastsequence            
               ->  Bitmap Index Scan on aotable_fastsequence_idx 
 Optimizer: Postgres query optimizer                             
(7 rows)
SELECT * FROM aotable_fastsequence WHERE i < 10 ORDER BY i;
 i 
---
 1 
 2 
 3 
 4 
 5 
 6 
 7 
 8 
 9 
(9 rows)

-- Verify table's gp_fastsequence value is preserved
SELECT fs.gp_segment_id, fs.objmod, fs.last_sequence FROM pg_class c JOIN pg_appendonly ao ON c.oid=ao.relid JOIN gp_dist_random('gp_fastsequence') fs ON ao.segrelid=fs.objid WHERE c.relname='aocotable_fastsequence' ORDER BY 1, 2, 3;
 gp_segment_id | objmod | last_sequence 
---------------+--------+---------------
 0             | 0      | 0             
 0             | 1      | 301           
 0             | 2      | 101           
 1             | 0      | 0             
 1             | 1      | 301           
 1             | 2      | 101           
 2             | 0      | 0             
 2             | 1      | 301           
 2             | 2      | 101           
(9 rows)

-- Verify table data is not corrupt using seqscan
SET enable_indexscan = false;
SET
SET enable_bitmapscan = false;
SET
SET enable_seqscan = true;
SET

EXPLAIN (COSTS OFF) SELECT * FROM aocotable_fastsequence ORDER BY i;
 QUERY PLAN                                     
------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)       
   Merge Key: i                                 
   ->  Sort                                     
         Sort Key: i                            
         ->  Seq Scan on aocotable_fastsequence 
 Optimizer: Postgres query optimizer            
(6 rows)
SELECT * FROM aocotable_fastsequence ORDER BY i;
 i   
-----
 1   
 2   
 3   
 4   
 5   
 6   
 7   
 8   
 9   
 10  
 11  
 12  
 13  
 14  
 15  
 16  
 17  
 18  
 19  
 20  
 21  
 22  
 23  
 24  
 25  
 26  
 27  
 28  
 29  
 30  
 102 
 103 
 104 
 105 
 106 
 107 
 108 
 109 
 110 
 111 
 112 
 113 
 114 
 115 
 116 
 117 
 118 
 119 
 120 
 121 
(50 rows)

-- Verify INSERTs produce no duplicate ctids
1: BEGIN;
BEGIN
1: INSERT INTO aocotable_fastsequence SELECT generate_series(1001, 1010);
INSERT 10
2: INSERT INTO aocotable_fastsequence SELECT generate_series(1011, 1020);
INSERT 10
1: COMMIT;
COMMIT
SELECT gp_segment_id, ctid, count(ctid) FROM aocotable_fastsequence GROUP BY gp_segment_id, ctid HAVING count(ctid) > 1;
 gp_segment_id | ctid | count 
---------------+------+-------
(0 rows)

SET enable_indexscan = true;
SET
SET enable_bitmapscan = true;
SET
SET enable_seqscan = false;
SET

EXPLAIN (COSTS OFF) SELECT * FROM aocotable_fastsequence ORDER BY i;
 QUERY PLAN                                                        
-------------------------------------------------------------------
 Gather Motion 3:1  (slice1; segments: 3)                          
   Merge Key: i                                                    
   ->  Sort                                                        
         Sort Key: i                                               
         ->  Bitmap Heap Scan on aocotable_fastsequence            
               ->  Bitmap Index Scan on aocotable_fastsequence_idx 
 Optimizer: Postgres query optimizer                               
(7 rows)
SELECT * FROM aocotable_fastsequence WHERE i < 10 ORDER BY i;
 i 
---
 1 
 2 
 3 
 4 
 5 
 6 
 7 
 8 
 9 
(9 rows)
